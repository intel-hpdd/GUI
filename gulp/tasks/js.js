//
// INTEL CONFIDENTIAL
//
// Copyright 2013-2016 Intel Corporation All Rights Reserved.
//
// The source code contained or described herein and all documents related
// to the source code ("Material") are owned by Intel Corporation or its
// suppliers or licensors. Title to the Material remains with Intel Corporation
// or its suppliers and licensors. The Material contains trade secrets and
// proprietary and confidential information of Intel or its suppliers and
// licensors. The Material is protected by worldwide copyright and trade secret
// laws and treaty provisions. No part of the Material may be used, copied,
// reproduced, modified, published, uploaded, posted, transmitted, distributed,
// or disclosed in any way without Intel's prior express written permission.
//
// No license under any patent, copyright, trade secret or other intellectual
// property right is granted to or conferred upon you by disclosure or delivery
// of the Materials, either expressly, by implication, inducement, estoppel or
// otherwise. Any license under such intellectual property rights must be
// express and approved by Intel in writing.

'use strict';

var paths = require('../paths.json');
var gulp = require('gulp');
var babel = require('gulp-babel');
var sourcemaps = require('gulp-sourcemaps');
var destDir = require('../dest-dir');
var relativeSourcemapsSource = require('gulp-relative-sourcemaps-source');

function jsDeps (fn) {
  return gulp.src(paths.js.deps, {
    since: gulp.lastRun(fn),
    base: '.'
  });
}

exports.jsDepsDev = function jsDepsDev () {
  return jsDeps(jsDepsDev)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

exports.jsDepsProd = function jsDepsProd () {
  return jsDeps(jsDepsProd)
  .pipe(gulp.dest('./dest'));
};

function socketWorker (fn) {
  return gulp.src(paths.js.socketWorker, {
    since: gulp.lastRun(fn)
  })
  .pipe(sourcemaps.init({ loadMaps: true }))
  .pipe(sourcemaps.write('.'));
}

exports.socketWorkerDev = function socketWorkerDev () {
  return socketWorker(socketWorkerDev)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

exports.socketWorkerProd = function socketWorkerProd () {
  return socketWorker(socketWorkerProd)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.dest('./dist'));
};

function systemConfig (fn) {
  return gulp.src(paths.js.config, {
    since: gulp.lastRun(fn),
    base: '.'
  });
}

exports.systemConfigDev = function systemConfigDev () {
  return systemConfig(systemConfigDev)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

var babelDev = babel.bind(null, {
  presets: [],
  babelrc: false,
  plugins: [
    'check-es2015-constants',
    'transform-flow-strip-types',
    'transform-class-properties',
    'transform-object-rest-spread',
    'transform-async-to-generator',
    'transform-es2015-classes',
    'transform-es2015-modules-systemjs'
  ]
});

var babelProd = babel.bind(null, {
  presets: ['es2015'],
  babelrc: false,
  plugins: [
    'transform-flow-strip-types',
    'transform-class-properties',
    'transform-object-rest-spread',
    [
      'angularjs-annotate',
      {
        explicitOnly: true
      }
    ],
    'transform-es2015-modules-systemjs'
  ]
});

function jsSource (fn) {
  return gulp.src(paths.js.source, {
    since: gulp.lastRun(fn),
    base: '.'
  })
  .pipe(sourcemaps.init());
}

exports.jsSourceDev = function jsSourceDev () {
  return jsSource(jsSourceDev)
  .pipe(babelDev())
  .pipe(sourcemaps.write({ sourceRoot: '/static/chroma_ui' }))
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

exports.jsSourceProd = function jsSourceProd () {
  return jsSource(jsSourceProd)
  .pipe(babelProd())
  .pipe(relativeSourcemapsSource({dest: 'dest'}))
  .pipe(sourcemaps.write({ sourceRoot: '' }))
  .pipe(gulp.dest('./dest'));
};

exports.jsTest = function jsTest () {
  return gulp.src(paths.js.tests, {
    since: gulp.lastRun(jsTest),
    base: '.'
  })
  .pipe(sourcemaps.init())
  .pipe(babelDev())
  .pipe(sourcemaps.write({ sourceRoot: '' }))
  .pipe(gulp.dest('./dest'));
};

exports.jsTestFixtures = function jsTestFixtures () {
  return gulp.src(paths.js.testFixtures, {
    since: gulp.lastRun(jsTestFixtures),
    base: '.'
  })
  .pipe(gulp.dest('./dest'));
};

exports.jsTestDeps = function jsTestDeps () {
  return gulp.src(paths.js.testDeps, {
    since: gulp.lastRun(jsTestDeps),
    base: '.'
  })
  .pipe(gulp.dest('./dest'));
};
