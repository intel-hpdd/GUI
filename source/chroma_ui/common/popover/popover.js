//
// INTEL CONFIDENTIAL
//
// Copyright 2013-2014 Intel Corporation All Rights Reserved.
//
// The source code contained or described herein and all documents related
// to the source code ("Material") are owned by Intel Corporation or its
// suppliers or licensors. Title to the Material remains with Intel Corporation
// or its suppliers and licensors. The Material contains trade secrets and
// proprietary and confidential information of Intel or its suppliers and
// licensors. The Material is protected by worldwide copyright and trade secret
// laws and treaty provisions. No part of the Material may be used, copied,
// reproduced, modified, published, uploaded, posted, transmitted, distributed,
// or disclosed in any way without Intel's prior express written permission.
//
// No license under any patent, copyright, trade secret or other intellectual
// property right is granted to or conferred upon you by disclosure or delivery
// of the Materials, either expressly, by implication, inducement, estoppel or
// otherwise. Any license under such intellectual property rights must be
// express and approved by Intel in writing.


angular.module('iml-popover', ['position']).directive('imlPopover', ['position', '$timeout', '$window', '$compile',
  function (position, $timeout, $window, $compile) {
    'use strict';

    var template = '<div class="popover fade {{ placement }}" ng-class="{in: open}">\
<div class="arrow"></div>\
<h3 class="popover-title" ng-bind="title" ng-show="title"></h3>\
<div class="popover-content"></div>\
</div>';

    return {
      restrict: 'E',
      transclude: true,
      scope: {
        placement: '@',
        title: '@',
        work: '&',
        onToggle: '&'
      },
      link: function link (scope, el, attrs, ctrl, $transclude) {
        var popoverLinker = $compile(template);

        var popoverButton = el.siblings('.activate-popover').eq(0),
          wrappedWindow = angular.element($window);

        if (!popoverButton) throw new Error('No popover button found.');

        scope.open = false;

        scope.work({
          actions: {
            hide: hide,
            recalculate: recalculate
          }
        });

        popoverButton.on('click', function handleClick ($event) {
          // Close any other popovers that are currently open
          if (!scope.open)
            wrappedWindow.trigger('click');

          toggle($event);
          scope.$digest();
        });

        scope.$on('$destroy', function onDestroy () {
          popoverButton.off('click');
          wrappedWindow.off('click', digestAndHide);
          wrappedWindow = popoverButton = popoverLinker = null;

          destroyPopover();
        });

        var inheritedScope, popoverEl, positioner;
        function createPopover () {
          $transclude(function createNewPopover (clone, transcludedScope) {
            inheritedScope = transcludedScope;

            popoverEl = popoverLinker(scope, angular.noop);

            positioner = position.positioner(popoverEl[0]);

            popoverEl.find('.popover-content').append(clone);
            popoverEl.on('click', function handleClick ($event) {
              $event.stopPropagation();
            });

            el.before(popoverEl);

            scope.onToggle({
              state: 'opened'
            });

            $timeout(function showAndRecalculate () {
              transcludedScope.$parent.$digest();
              popoverEl.css('display', 'block');

              if (scope.placement)
                recalculate();

              scope.open = true;
              scope.$digest();
            }, 0, false);
          });
        }

        function destroyPopover () {
          scope.onToggle({
            state: 'closed'
          });

          if (popoverEl) {
            popoverEl.off('click');
            popoverEl.remove();
            popoverEl = null;
          }

          if (inheritedScope) {
            inheritedScope.$destroy();
            inheritedScope = null;
          }
        }

        /**
         * Toggles the visibility of the popover. Used as an event callback.
         * @param {object} $event
         */
        function toggle ($event) {
          $event.stopPropagation();

          if (scope.open) {
            hide();
          } else {
            createPopover();
            wrappedWindow.on('click', digestAndHide);
          }
        }

        function digestAndHide () {
          hide();
          scope.$digest();
        }

        /**
         * destroys the popover and unbinds the window click listener.
         */
        function hide () {
          scope.open = false;

          $timeout(destroyPopover, 500);

          wrappedWindow.off('click', digestAndHide);
        }

        /**
         * Figures out the placement of the popover and applies it to the element's style.
         */
        function recalculate () {
          if (!popoverEl)
            return;

          popoverEl.css('min-width', '');
          popoverEl.css(position.position(scope.placement, positioner));
        }
      }
    };
  }
]);
