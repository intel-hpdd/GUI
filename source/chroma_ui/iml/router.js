//
// INTEL CONFIDENTIAL
//
// Copyright 2013-2015 Intel Corporation All Rights Reserved.
//
// The source code contained or described herein and all documents related
// to the source code ("Material") are owned by Intel Corporation or its
// suppliers or licensors. Title to the Material remains with Intel Corporation
// or its suppliers and licensors. The Material contains trade secrets and
// proprietary and confidential information of Intel or its suppliers and
// licensors. The Material is protected by worldwide copyright and trade secret
// laws and treaty provisions. No part of the Material may be used, copied,
// reproduced, modified, published, uploaded, posted, transmitted, distributed,
// or disclosed in any way without Intel's prior express written permission.
//
// No license under any patent, copyright, trade secret or other intellectual
// property right is granted to or conferred upon you by disclosure or delivery
// of the Materials, either expressly, by implication, inducement, estoppel or
// otherwise. Any license under such intellectual property rights must be
// express and approved by Intel in writing.


angular.module('imlRoutes', ['imlRouterModule',
  'lnetModule', 'dashboard', 'server', 'statusModule'])
  .config(['$routeSegmentProvider', 'GROUPS', function ($routeSegmentProvider, GROUPS) {
    $routeSegmentProvider
      .within('app')
      .segment('dashboard', untilResolved({
        controller: 'DashboardCtrl',
        controllerAs: 'dashboard',
        templateUrl: 'iml/dashboard/assets/html/dashboard.html',
        resolve: {
          fsStream: ['dashboardFsStream', fp.invoke(fp.__, [])],
          hostStream: ['dashboardHostStream', fp.invoke(fp.__, [])],
          targetStream: ['dashboardTargetStream', fp.invoke(fp.__, [])]
        }
      }));

    $routeSegmentProvider
      .when('/', 'app.dashboard.base')
      .when('/dashboard', 'app.dashboard.base')
      .when('/dashboard/fs/:fsId', 'app.dashboard.base')
      .within('app')
      .within('dashboard')
      .segment('base', untilResolved({
        controller: 'BaseDashboardCtrl',
        controllerAs: 'baseDashboard',
        templateUrl: 'iml/dashboard/assets/html/base-dashboard.html',
        resolve: {
          charts: ['baseDashboardChartResolves', fp.invoke(fp.__, [])],
          fsStream: ['baseDashboardFsStream', fp.invoke(fp.__, [])]
        },
        dependencies: ['fsId']
      }));

    $routeSegmentProvider
      .when('/dashboard/server/:serverId', 'app.dashboard.server')
      .within('app')
      .within('dashboard')
      .segment('server', untilResolved({
        controller: 'ServerDashboardCtrl',
        controllerAs: 'serverDashboard',
        templateUrl: 'iml/dashboard/assets/html/server-dashboard.html',
        resolve: {
          charts: ['$route', 'serverDashboardChartResolves', function getCharts ($route, serverDashboardChartResolves) {
            return serverDashboardChartResolves($route.current.params.serverId);
          }],
          hostStream: ['$route', 'resolveStream', 'socketStream',
            function hostStream ($route, resolveStream, socketStream) {
              return resolveStream(socketStream('/host/' + $route.current.params.serverId, {
                jsonMask: 'label'
              }));
            }
          ]
        },
        dependencies: ['serverId']
      }));

    $routeSegmentProvider
      .when('/dashboard/fs/:fsId/MDT/:targetId', 'app.dashboard.target', { kind: 'MDT' })
      .when('/dashboard/server/:serverId/MDT/:targetId', 'app.dashboard.target', { kind: 'MDT' })
      .when('/dashboard/fs/:fsId/OST/:targetId', 'app.dashboard.target', { kind: 'OST' })
      .when('/dashboard/server/:serverId/OST/:targetId', 'app.dashboard.target', { kind: 'OST' })
      .within('app')
      .within('dashboard')
      .segment('target', untilResolved({
        controller: 'TargetDashboardController',
        controllerAs: 'targetDashboard',
        templateUrl: 'iml/dashboard/assets/html/target-dashboard.html',
        resolve: {
          kind: ['targetDashboardKind', fp.invoke(fp.__, [])],
          charts: ['targetDashboardResolves', fp.invoke(fp.__, [])],
          targetStream: ['targetDashboardTargetStream', fp.invoke(fp.__, [])],
          usageStream: ['targetDashboardUsageStream', fp.invoke(fp.__, [])]
        },
        dependencies: ['fsId', 'serverId', 'targetId']
      }));

    $routeSegmentProvider
      .when('/configure/server', 'app.server')
      .within('app')
      .segmentAuthenticated('server', untilResolved({
        controller: 'ServerCtrl',
        templateUrl: 'iml/server/assets/html/server.html',
        resolve: {
          streams: ['serverStreamsResolves', fp.invoke(fp.__, [])]
        },
        access: GROUPS.FS_ADMINS
      }));


    $routeSegmentProvider
      .when('/configure/server/:id', 'app.serverDetail')
      .within('app')
      .segmentAuthenticated('serverDetail', untilResolved({
        controller: 'ServerDetailController',
        controllerAs: 'serverDetail',
        templateUrl: 'iml/server/assets/html/server-detail.html',
        resolve: {
          streams: ['serverDetailResolves', '$route', function getStreams (serverDetailResolves, $route) {
            return serverDetailResolves($route);
          }]
        },
        access: GROUPS.FS_ADMINS
      }));

    $routeSegmentProvider.when('/dashboard/jobstats/:id/:startDate/:endDate', 'app.jobstats');

    $routeSegmentProvider
      .within('app')
      .segment('jobstats', untilResolved({
        controller: 'JobStatsCtrl',
        controllerAs: 'jobStats',
        templateUrl: 'iml/job-stats/assets/html/job-stats.html',
        resolve: {
          target: ['$route', 'TargetModel', function resolveTarget ($route, TargetModel) {
            return TargetModel.get({
              id: $route.current.params.id
            }).$promise;
          }],
          metrics: ['$q', '$route', 'TargetMetricModel', function resolveMetrics ($q, $route, TargetMetricModel) {
            var commonParams = {
              begin: $route.current.params.startDate,
              end: $route.current.params.endDate,
              job: 'id',
              id: $route.current.params.id
            };
            var metrics = ['read_bytes', 'write_bytes', 'read_iops', 'write_iops'];

            var promises = metrics.reduce(function reducer (out, metric) {

              var params = _.extend({}, commonParams, {metrics: metric});

              out[metric] = TargetMetricModel.getJobAverage(params);

              return out;
            }, {});

            return $q.all(promises);
          }]
        }
      }));

    $routeSegmentProvider
      .when('/status', 'app.status')
      .within('app')
      .segment('status', untilResolved({
        controller: 'StatusController',
        controllerAs: 'ctrl',
        templateUrl: 'iml/status/assets/html/status.html'
      }));

    function untilResolved (obj) {
      return _.extend({}, obj, {
        untilResolved: {
          templateUrl: 'common/loading/assets/html/loading.html'
        }
      });
    }
  }]);
