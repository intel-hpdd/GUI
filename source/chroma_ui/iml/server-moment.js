//
// INTEL CONFIDENTIAL
//
// Copyright 2013 Intel Corporation All Rights Reserved.
//
// The source code contained or described herein and all documents related
// to the source code ("Material") are owned by Intel Corporation or its
// suppliers or licensors. Title to the Material remains with Intel Corporation
// or its suppliers and licensors. The Material contains trade secrets and
// proprietary and confidential information of Intel or its suppliers and
// licensors. The Material is protected by worldwide copyright and trade secret
// laws and treaty provisions. No part of the Material may be used, copied,
// reproduced, modified, published, uploaded, posted, transmitted, distributed,
// or disclosed in any way without Intel's prior express written permission.
//
// No license under any patent, copyright, trade secret or other intellectual
// property right is granted to or conferred upon you by disclosure or delivery
// of the Materials, either expressly, by implication, inducement, estoppel or
// otherwise. Any license under such intellectual property rights must be
// express and approved by Intel in writing.


(function () {
  'use strict';

  angular.module('serverMoment', ['moment', 'environment'])
    .factory('getServerMoment', ['moment', 'SERVER_TIME_DIFF', getServerMomentFactory]);

  /**
   * Returns a getServerMoment function.
   * @param {Object} moment
   * @param {String} SERVER_TIME_DIFF
   * @returns {Function}
   */
  function getServerMomentFactory(moment, SERVER_TIME_DIFF) {
    /**
     * Creates a new moment with forwarded arguments.
     * The moment is then updated with SERVER_TIME_DIFF
     * and returned.
     * @returns {Object}
     */
    return function getServerMoment() {
      return moment
        .apply(moment, arguments)
        .add(SERVER_TIME_DIFF);
    };
  }
}());
