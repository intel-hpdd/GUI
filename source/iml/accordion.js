// @flow

//
// Copyright (c) 2018 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

import { Component } from "inferno";
import { cloneVNode } from "inferno-clone-vnode";

type PanelComponentPropsT = {
  collapsed: boolean,
  accordionId: string,
  panelId: string,
  headingId: string,
  children: React$Element<*>[]
};

type PanelComponentStateT = {
  collapsed: boolean
};

export class PanelComponent extends Component {
  state: PanelComponentStateT;
  props: PanelComponentPropsT;

  constructor(props: PanelComponentPropsT) {
    super(props);

    this.state = {
      collapsed: props.collapsed
    };
  }

  togglePanel() {
    this.setState({
      collapsed: !this.state.collapsed
    });
  }

  render() {
    const title = cloneVNode(this.props.children[0], { collapsed: this.state.collapsed });

    return (
      <div class="panel panel-default">
        <div class="panel-heading" role="tab" id={this.props.headingId}>
          <h4 class="panel-title">
            <a
              className={this.state.collapsed ? "collapsed" : ""}
              role="button"
              data-toggle="collapse"
              data-parent={`#${this.props.accordionId}`}
              aria-expanded={!this.props.collapsed}
              aria-controls={this.props.panelId}
              onClick={this.togglePanel.bind(this)}
            >
              {title}
            </a>
          </h4>
        </div>
        <div
          id={this.props.panelId}
          className={`panel-collapse collapse ${!this.state.collapsed ? "in" : ""}`}
          role="tabpanel"
          aria-labelledby={this.props.headingId}
        >
          <div class="panel-body">{this.props.children[1]}</div>
        </div>
      </div>
    );
  }
}

type AccordionT = {
  id: string,
  children: React$Element<*>
};

export default ({ id, children }: AccordionT) => {
  return (
    <div class="panel-group" id={id} role="tablist" aria-multiselectable="true">
      {children}
    </div>
  );
};
