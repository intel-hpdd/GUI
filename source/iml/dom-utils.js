// @flow

type QuerySelector<T: HTMLElement> = (HTMLElement | Document, string) => T;

export const querySelector: QuerySelector<*> = (
  elm: HTMLElement | Document,
  sel: string
) => {
  const selected = elm.querySelector(sel);

  if (!selected)
    throw new Error(
      `Could not find selector ${sel} for element ${elm.nodeName}`
    );

  return selected;
};
