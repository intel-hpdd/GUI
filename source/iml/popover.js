// @flow

//
// Copyright (c) 2017 Intel Corporation. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

import Inferno from 'inferno';
import Component from 'inferno-component';

type PopoverChildProps = {
  children: React$Element<*>[]
};

export class PopoverContainer extends Component {
  windowListener: ?Function;
  state: { isOpen: boolean };
  constructor(props: PopoverChildProps) {
    super(props);
    this.state = { isOpen: false };
  }
  windowHandler() {
    const { isOpen: previousIsOpen } = this.state;
    const isOpen = !previousIsOpen;

    this.setState({ isOpen });

    if (isOpen || !this.windowListener) return;

    window.removeEventListener('click', this.windowListener, false);
    this.windowListener = null;
  }
  handleClick() {
    if (this.windowListener) return;

    this.windowListener = this.windowHandler.bind(this);
    window.addEventListener('click', this.windowListener);
  }
  render() {
    const children = this.props.children.map(c => {
      if (!c.props) return c;
      else if (c.props.popoverButton)
        return Inferno.cloneVNode(c, {
          onClick: this.handleClick.bind(this)
        });
      else if (c.props.popover)
        return Inferno.cloneVNode(c, {
          visible: this.state.isOpen
        });
      else return c;
    });

    return (
      <span style="position: relative;">
        {children}
      </span>
    );
  }
}

export const PopoverTitle = ({ children }: PopoverChildProps) =>
  <h3 class="popover-title">
    {children}
  </h3>;

export const PopoverContent = ({ children }: PopoverChildProps) =>
  <div class="popover-content">
    {children}
  </div>;

type PopoverProps = {
  children?: React$Element<*>,
  visible?: boolean,
  direction: 'top' | 'bottom' | 'left' | 'right'
};

export const Popover = ({ children, visible, direction }: PopoverProps) => {
  if (!visible) return;

  return (
    <div className={`fade popover ${direction} in`}>
      <div class="arrow" />
      {children}
    </div>
  );
};
