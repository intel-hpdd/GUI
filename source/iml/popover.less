.popover-container {
  position: relative;

  &.popover-hover:hover .popover {
    display: block;
    opacity: 0.9;
    max-width: 200px;

    .popover-inner {
      word-wrap: break-word;
    }
  }
}

.popover {
  top: auto;
  left: auto;
  width: 200px;

  &.xsmall {
    width: 50px;
  }

  &.small {
    width: 100px;
  }

  &.medium {
    width: 150px;
  }

  &.large {
    width: 200px;
  }

  &.right {
    left: 100%;
    bottom: 50%;
    transform: translateY(50%);
  }

  &.left {
    bottom: 50%;
    transform: translate(-100%, 50%);
  }

  &.bottom {
    transform: translateX(50%);
    top: 100%;
    right: 50%;
  }

  &.top {
    transform: translate(50%, -100%);
    top: 0;
    right: 50%;
  }
}

.popover {
  display: none;

  &.in {
    display: block;
  }
}
