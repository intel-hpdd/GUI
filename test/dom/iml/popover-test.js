// @flow

import { render } from "inferno";
import { Popover, PopoverContent, PopoverTitle, PopoverContainer } from "../../../source/iml/popover";
import WindowClickListener from "../../../source/iml/window-click-listener.js";
import { renderToSnapshot } from "../../test-utils.js";
import { querySelector } from "../../../source/iml/dom-utils.js";

describe("Popover DOM testing", () => {
  it("should render correctly without a title", () => {
    expect(
      renderToSnapshot(
        <Popover visible={true} direction="bottom">
          <PopoverContent>Foo</PopoverContent>
        </Popover>
      )
    ).toMatchSnapshot();
  });

  it("should render correctly with a title", () => {
    expect(
      renderToSnapshot(
        <Popover visible={true} direction="bottom">
          <PopoverTitle>Hi</PopoverTitle>
          <PopoverContent>Foo</PopoverContent>
        </Popover>
      )
    ).toMatchSnapshot();
  });

  it("should render sub-props correctly", () => {
    const message = "This is some content";

    expect(
      renderToSnapshot(
        <Popover visible={true} direction="right">
          <PopoverTitle>Hi</PopoverTitle>
          <PopoverContent>{message}</PopoverContent>
        </Popover>
      )
    ).toMatchSnapshot();
  });

  it("should not render if not visible", () => {
    expect(
      renderToSnapshot(
        <Popover visible={false} direction="right">
          <PopoverContent>bar</PopoverContent>
        </Popover>
      )
    ).toMatchSnapshot();
  });

  describe("Container", () => {
    let root, vNode;

    describe("with popover button", () => {
      describe("containing no children", () => {
        beforeEach(() => {
          root = document.createElement("div");

          vNode = (
            <WindowClickListener>
              <PopoverContainer>
                <button popoverButton={true} />
                <Popover popover={true} direction="right">
                  <PopoverContent>bar</PopoverContent>
                </Popover>
              </PopoverContainer>
            </WindowClickListener>
          );

          render(vNode, root);

          if (document.body) document.body.appendChild(root);
        });

        afterEach(() => {
          render(null, root);
          if (document.body) document.body.removeChild(root);
        });

        it("should render correctly", () => {
          expect(renderToSnapshot(vNode)).toMatchSnapshot();
        });

        it("should open and close when clicking the button", () => {
          expect.assertions(2);

          querySelector(root, "button").click();

          expect(root.innerHTML).toMatchSnapshot();

          querySelector(root, "button").click();

          expect(root.innerHTML).toMatchSnapshot();
        });
      });

      describe("containing children", () => {
        beforeEach(() => {
          root = document.createElement("div");

          vNode = (
            <WindowClickListener>
              <PopoverContainer>
                <span popoverButton={true}>
                  <i />
                  <Popover popover={true} direction="bottom">
                    <PopoverTitle>Alerts</PopoverTitle>
                    <PopoverContent>
                      <ul>
                        <li key={1}>message</li>
                      </ul>
                    </PopoverContent>
                  </Popover>
                  <span>Some other element</span>
                </span>
                <></>
                <div dummy-param={"dummy data"} />
              </PopoverContainer>
            </WindowClickListener>
          );

          render(vNode, root);

          if (document.body) document.body.appendChild(root);
        });

        afterEach(() => {
          render(null, root);
          if (document.body) document.body.removeChild(root);
        });

        it("should match snapshot", () => {
          expect(root).toMatchSnapshot();
        });
      });
    });
  });
});
