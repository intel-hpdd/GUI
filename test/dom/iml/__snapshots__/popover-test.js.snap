// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover DOM testing Container should open and close when clicking the button 1`] = `"<span style=\\"position: relative;\\"><button popoverbutton=\\"true\\"></button><div class=\\"fade popover right in\\"><div class=\\"arrow\\"></div><div class=\\"popover-content\\">bar</div></div></span>"`;

exports[`Popover DOM testing Container should open and close when clicking the button 2`] = `"<span style=\\"position: relative;\\"><button popoverbutton=\\"true\\"></button></span>"`;

exports[`Popover DOM testing Container should render correctly 1`] = `"<span style=\\"position: relative;\\"><button popoverbutton=\\"true\\"></button></span>"`;

exports[`Popover DOM testing should not render if not visible 1`] = `""`;

exports[`Popover DOM testing should render correctly with a title 1`] = `"<div class=\\"fade popover bottom in\\"><div class=\\"arrow\\"></div><h3 class=\\"popover-title\\">Hi</h3><div class=\\"popover-content\\">Foo</div></div>"`;

exports[`Popover DOM testing should render correctly without a title 1`] = `"<div class=\\"fade popover bottom in\\"><div class=\\"arrow\\"></div><div class=\\"popover-content\\">Foo</div></div>"`;

exports[`Popover DOM testing should render sub-props correctly 1`] = `"<div class=\\"fade popover right in\\"><div class=\\"arrow\\"></div><h3 class=\\"popover-title\\">Hi</h3><div class=\\"popover-content\\">This is some content</div></div>"`;
